internal/authentication/ldap_client_factory.go

```diff
@@ -117,162 +121,393 @@ type PooledLDAPClientFactory struct {
 	if len(opts) != 0 {
 		return getLDAPClient(f.config.Address.String(), f.config.User, f.config.Password, f.config.Timeout, f.dialer, f.tls, f.config.StartTLS, f.opts, opts...)
 	}
 
-	return f.acquire(context.Background())
+	return f.acquire()
 }
 
+// TODO: Clarify why this method is not thread-safe - or does this only apply to the returnd client?
+
 // The new function creates a pool based client. This function is not thread safe.
+// NOTE: this function adheres to config.Timeout, not config.Pooling.Timeout!
 func (f *PooledLDAPClientFactory) new() (pooled *LDAPClientPooled, err error) {
-	var client ldap.Client
-
-	if client, err = getLDAPClient(f.config.Address.String(), f.config.User, f.config.Password, f.config.Timeout, f.dialer, f.tls, f.config.StartTLS, f.opts); err != nil {
-		return nil, fmt.Errorf("error occurred establishing new client for the pool: %w", err)
-	}
+	client, err := getLDAPClient(f.config.Address.String(), f.config.User, f.config.Password, f.config.Timeout, f.dialer, f.tls, f.config.StartTLS, f.opts)
+	if err != nil { return nil, err }
 
 	return &LDAPClientPooled{Client: client}, nil
 }
 
-// ReleaseClient returns a client using the pool or closes it.
+// thread-safe, returns a client using the pool or closes it.
 func (f *PooledLDAPClientFactory) ReleaseClient(client ldap.Client) (err error) {
-	f.mu.Lock()
-
-	defer f.mu.Unlock()
+	logging.Logger().Trace("Releasing LDAP client")
 
-	if f.closing {
+	pooled, ok := client.(*LDAPClientPooled)
+	if !ok {
+		logging.Logger().Debug("LDAP client is not pooled, closing directly")
 		return client.Close()
 	}
 
-	if pool, ok := client.(*LDAPClientPooled); !ok || cap(f.pool) == len(f.pool) {
-		// Prevent extra or non-pool connections from being returned into the pool.
-		return client.Close()
-	} else {
-		f.pool <- pool
+	if !pooled.IsHealthy() || pooled.IsExpired() {
+		logging.Logger().Info("Pooled LDAP client is not healthy or has expired, disposing")
+		return f.disposeClient(pooled)
 	}
 
-	return nil
+	pC := f.pool
+	if pC == nil {
+		logging.Logger().Warning("LDAP pool is not initialized or was closed, disposing client")
+		return f.disposeClient(pooled)
+	}
+
+	// NOTE: clients can still be returned while the pool is closing.
+	// Rationale: cleanup in poolManager() handles this gracefully.
+
+	select {
+	case pC <- pooled:
+		logging.Logger().Trace("Successfully returned client to LDAP pool")
+		return nil
+	default: // shouldn't happen, unless returning more clients than acquired
+		logging.Logger().Debug("LDAP pool is full, disposing client")
+		return f.disposeClient(pooled)
+	}
 }
 
-func (f *PooledLDAPClientFactory) acquire(ctx context.Context) (client *LDAPClientPooled, err error) {
-	f.mu.Lock()
+// thread-safe, closes the client and decrements active count
+func (f *PooledLDAPClientFactory) disposeClient(client *LDAPClientPooled) error {
+	if client == nil {
+		logging.Logger().Warn("Cannot dispose nil LDAP client")
```

Comment:
Please remove call to `logging.Logger()` and initialize a fielded logger like in #10383